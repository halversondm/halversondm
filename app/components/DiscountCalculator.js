/**
 * Created by Daniel on 6/26/2016.
 */
"use strict";

import React from "react";

const DiscountCalculator = React.createClass({
  render() {
    return <div>
    </div>;
  }
});

export default DiscountCalculator;
