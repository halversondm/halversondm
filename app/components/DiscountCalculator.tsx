/**
 * Created by Daniel on 6/26/2016.
 */
"use strict";

import * as React from "react";
import DiscountCalculatorService from "./DiscountCalculatorService";

interface DiscountCalculatorState {
    labelPrice: string,
    discount1: string,
    discount2: string,
    errorMessage: string[],
    successMessage: string
}

export default class DiscountCalculator extends React.Component<undefined,DiscountCalculatorState> {

    discountCalculatorService: DiscountCalculatorService;
    state: DiscountCalculatorState;

    constructor() {
        super();
        this.state = this.initialState();
        this.discountCalculatorService = new DiscountCalculatorService();
        this.calculate = this.calculate.bind(this);
        this.clear = this.clear.bind(this);
        this.labelPriceChange = this.labelPriceChange.bind(this);
        this.discount1Change = this.discount1Change.bind(this);
        this.discount2Change = this.discount2Change.bind(this);
    }

    private initialState() {
        return {
            labelPrice: "",
            discount1: "",
            discount2: "",
            errorMessage: [],
            successMessage: ""
        };
    }

    calculate() {
        this.discountCalculatorService.validate(this.state.discount1, this.state.discount2, this.state.labelPrice);
        if (this.discountCalculatorService.isError()) {
            this.setState({
                errorMessage: this.discountCalculatorService.getMessage(),
                successMessage: ""
            });
        } else {
            this.discountCalculatorService.calculate();
            this.setState({
                successMessage: this.discountCalculatorService.getMessage()[0],
                errorMessage: []
            });
        }
    }

    clear() {
        this.setState(this.initialState());
    }

    labelPriceChange(event: any) {
        this.setState({labelPrice: event.target.value});
    }

    discount1Change(event: any) {
        this.setState({discount1: event.target.value});
    }

    discount2Change(event: any) {
        this.setState({discount2: event.target.value});
    }

    render() {
        return <div>
            <h2 className="text-primary">Discount Calculator</h2>
            <p>Calculate your final price where the retailer is offering separate
                discounts.</p>
            <fieldset>
                <form className="form-horizontal" role="form">
                    <div className="form-group">
                        <label htmlFor="labelPrice" className="col-sm-2 control-label">Label
                            Price
                        </label>
                        <div className="col-sm-10">
                            <div className="input-group">
                                <span className="input-group-addon">$</span>
                                <input type="number" className="form-control" id="labelPrice"
                                       onChange={this.labelPriceChange}
                                       value={this.state.labelPrice}/>
                            </div>
                        </div>
                    </div>
                    <div className="form-group">
                        <label htmlFor="discount1" className="col-sm-2 control-label">Discount
                            #1
                        </label>
                        <div className="col-sm-10">
                            <div className="input-group">
                                <input type="number" className="form-control" id="discount1"
                                       onChange={this.discount1Change}
                                       value={this.state.discount1}/> <span
                                className="input-group-addon">%</span>
                            </div>
                        </div>
                    </div>
                    <div className="form-group">
                        <label htmlFor="discount2" className="col-sm-2 control-label">Discount
                            #2
                        </label>
                        <div className="col-sm-10">
                            <div className="input-group">
                                <input type="number" className="form-control" id="discount2"
                                       onChange={this.discount2Change}
                                       value={this.state.discount2}/> <span
                                className="input-group-addon">%</span>
                            </div>
                        </div>
                    </div>
                    <div className="form-group">
                        <div className="col-sm-offset-2 col-sm-10">
                            <button type="button" id="calculate"
                                    className="btn btn-primary"
                                    onClick={this.calculate}>Calculate
                            </button>
                            <button type="button" id="reset"
                                    className="btn btn-danger"
                                    onClick={this.clear}>Reset
                            </button>
                        </div>
                    </div>
                    <div className="alert alert-danger alert-dismissible" role="alert"
                         hidden={this.state.errorMessage.length === 0}>
                        <button type="button" className="close" data-dismiss="alert"><span
                            aria-hidden="true">&times;</span></button>
                        <strong>Error! </strong>
                        <ul>
                            {
                                this.state.errorMessage.map((message, i) => {
                                    return <li key={i}>{message}</li>;
                                })
                            }
                        </ul>
                    </div>
                    <div className="alert alert-success alert-dismissible" role="alert"
                         hidden={this.state.successMessage === ""}>
                        <button type="button" className="close" data-dismiss="alert"><span
                            aria-hidden="true">&times;</span></button>
                        {this.state.successMessage}
                    </div>
                </form>
            </fieldset>

        </div>;
    }
}
