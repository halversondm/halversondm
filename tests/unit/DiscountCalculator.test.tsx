import React from "react";
import { render, screen } from '@testing-library/react';
import DiscountCalculator from "../../app/components/DiscountCalculator";
import userEvent from "@testing-library/user-event";

describe("Discount Calculator", () =>{
    it("Will render the component for a snapshot", () => {
        const component = render(<DiscountCalculator />);
        expect(component.container.toString()).toMatchSnapshot();
    });

    it("Will test a discount with $100, 80% first and 20% second to equal $16", async () => {
        render(<DiscountCalculator />);
        await userEvent.type(document.getElementById("labelPrice"), '100');
        await userEvent.type(document.getElementById("discount1"), '80');
        await userEvent.type(document.getElementById("discount2"), '20');
        await userEvent.click(document.getElementById("calculate"));

        expect(screen.getByText("Your final price is $16.00 plus tax")).not.toBeNull();
    });

    it("Will test a discount with $100, 80% first and 0% second to equal $20", async () => {
        render(<DiscountCalculator />);
        await userEvent.type(document.getElementById("labelPrice"), '100');
        await userEvent.type(document.getElementById("discount1"), '80');
        await userEvent.click(document.getElementById("calculate"));

        expect(screen.getByText("Your final price is $20.00 plus tax")).not.toBeNull();
    });

    it("Will test a discount with $100 and no discounts", async () => {
        render(<DiscountCalculator />);
        await userEvent.type(document.getElementById("labelPrice"), '100');
        await userEvent.click(document.getElementById("calculate"));

        expect(screen.getByText("Discount #1 is required and must be a number")).not.toBeNull();
    });

    it("Will with no input", async () => {
        render(<DiscountCalculator />);
        await userEvent.click(document.getElementById("calculate"));

        expect(screen.getByText("Label price is required and must be a number")).not.toBeNull();
        expect(screen.getByText("Discount #1 is required and must be a number")).not.toBeNull();
    });
})
