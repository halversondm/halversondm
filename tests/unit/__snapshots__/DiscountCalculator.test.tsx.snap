// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Discount Calculator > Will render the component for a snapshot 1`] = `"<div><div><h2 class="text-primary">Discount Calculator</h2><p>Calculate your final price where the retailer is offering separate discounts.</p><fieldset><form class="form-horizontal" role="form"><div class="form-group"><label for="labelPrice" class="col-sm-2 control-label">Label Price</label><div class="col-sm-10"><div class="input-group"><span class="input-group-addon">$</span><input class="form-control" id="labelPrice" type="number" value=""></div></div></div><div class="form-group"><label for="discount1" class="col-sm-2 control-label">Discount #1</label><div class="col-sm-10"><div class="input-group"><input class="form-control" id="discount1" type="number" value=""> <span class="input-group-addon">%</span></div></div></div><div class="form-group"><label for="discount2" class="col-sm-2 control-label">Discount #2</label><div class="col-sm-10"><div class="input-group"><input class="form-control" id="discount2" type="number" value=""> <span class="input-group-addon">%</span></div></div></div><div class="form-group"><div class="col-sm-offset-2 col-sm-10"><button type="button" id="calculate" class="btn btn-primary">Calculate</button><button type="button" id="reset" class="btn btn-danger">Reset</button></div></div><div class="alert alert-danger alert-dismissible" role="alert" id="errorMessages" hidden=""><button type="button" class="close" data-dismiss="alert"><span aria-hidden="true">×</span></button><strong>Error! </strong><ul></ul></div><div class="alert alert-success alert-dismissible" role="alert" id="successMessage" hidden=""><button type="button" class="close" data-dismiss="alert"><span aria-hidden="true">×</span></button></div></form></fieldset></div></div>"`;
